### ISSUE TITLE
Create new Pull Request on demand

### ISSUE THREAD

**Issue by mabreuortega** [cake-contrib/Cake.Issues.PullRequests.Tfs#7]
I'm doing this now, instead of adding the comments to an existing PR, I like to create a new PR to address the issues found.

```
var pullRequestSettings =
         new TfsPullRequestSettings(
            new Uri("http://myserver:8080/tfs/defaultcollection/myproject/_git/myrepository"),
            "refs/heads/feature/myfeature",
             TfsAuthenticationNtlm());

TfsCreatePullRequest(
         pullRequestSettings,
         "Title and more parameters");
```

Best regards,
Miguel

**Comment by pascalberger** [comment]
@mabreuortega What do you expect if there's is already a pull request for the branch? Do you just want to create a pull request if there's none existing yet?

**Comment by mabreuortega** [comment]
Both scenarios are valid:
* You want to solve StyleCop or Warning issues in a different PR and not now. You can not always have code debt time allocation an need to plan to do it in the future.

**Comment by pascalberger** [comment]
@mabreuortega If I understand correctly you wan't to use pull requests to manage technical dept. IMHO this doesn't make much sense. The purpose of a pull request is to request to merge a source branch into a target branch. Having multiple pull requests for the same source branch doesn't make much sense nor is it possible. So what would you expect if there is already an existing pull request?

If you're after getting just a list of issues, you can a `Cake.Issues.Reporting` addin (like [Cake.Issues.Reporting.Generic](https://github.com/cake-contrib/Cake.Issues.Reporting.Generic) which is currently in beta). If you want to manage technical dept a tool like [SonarQube](https://www.sonarqube.org/) is a much better fit.

**Comment by mabreuortega** [comment]
I wanted more to create PR to fix code quality issues later because the
main functionality needs to be ready now (real world things).
Anyway, another use case can be to create automatic merge PRs.
For example, when using GitFlow you need to merge automatically the master
changes to the develop branch, instead of doing a merge that may have
conflict, create a PR from master to develop assigned to the Lead Developer
or auto-approved (this give you the possibility to revert the PR in TFS
2017 )

**Comment by pascalberger** [comment]
I'm thinking about splitting the generic TFS part of this addin into a separate Cake.TFS addin which would have no dependencies to `Cake.Issues`. The drawback of this approach would be that it adds another dependency to this addin and thus would complicate dependency / versioning management for this addin.

**Comment by mabreuortega** [comment]
I already did that, check this out:
https://github.com/mabreuortega/Cake.Tfs

**Comment by pascalberger** [comment]
@mabreuortega Nice! Are you interested in moving the repository into the cake-contrib organization to simplify contributions by others? @gep13 can give you instructions

**Comment by mabreuortega** [comment]
I already did that and is listed under the TFS category.

**Comment by pascalberger** [comment]
I meant moving the repository to the [cake-contrib GitHub organization](https://github.com/cake-contrib/) (and ideally also use similar workflow, build process, etc as other addins)?

Is there a reason why the addin is called `Cake.Tfs.AutoMerge` and not just `Cake.Tfs`, as this limits the possible features which fit into the addin. E.g. the authentication code should be usable for much more than just working with pull requests.

I also saw that you currently don't have any license for your code. You should add a license for your code, as it looks like you've copied some code from this repository it should be compatible with the [license](https://github.com/cake-contrib/Cake.Issues.PullRequests.Tfs/blob/develop/LICENSE) from this repository (basically give proper attribution, especially as this repository also contains code inspired from code written by Microsoft and Sonar Source, probably not the parts which you've copied).

**Comment by mabreuortega** [comment]
I have no problem at all to move the repo to the Cake Contrib organization.

My original attempt was to create AutoMerge pull request only, the moment in time another idea appears we can refactor the auth and the connection factory logic to a Common library, and keep each package individually scoped and isolated.

Done with the licensing issue. I extract from here authentication and connection creation part.

**Comment by gep13** [comment]
@mabreuortega if you are interested in transferring the repository into the cake-contrib Organisation, the process is documented here:

https://github.com/cake-contrib/Home#how-do-i-as-an-addin-author-transfer-a-repository-to-the-organisation

It is really only the first part that you need to take of, i.e. transferring the repository to myself on GitHub.  Once that is done, I can take care of the rest of it.

**Comment by pascalberger** [comment]
@mabreuortega My idea for the different plugins was something like this:

- Cake.Tfs: Provides aliases for working with TFS, but also aliases for the different authentication types. 
- Cake.Issues: Provides base classes for reading issues
- Cake.Issues.PullRequests: Provides logic to determine issues which should be posted to pull requests. Requires Cake.Issues.
- Cake.Issues.PullRequests.Tfs: Implementation for posting issues to TFS pull requests. Requires Cake.Issues, Cake.Issues.PullRequests and the authentication aliases from Cake.Tfs (to avoid duplicated aliases).

To further avoid duplicated code it can make sense to put some TFS related code (like the authentication classes, or connection stuff) into a separate NuGet library (without any relation to Cake), which can be shared between the different Cake addins. The drawback of this is that it adds another dependency which needs to be merged into the assembly (which is not that easy for .NET core as for full framework with Costura.Fody).
